source/common/stats/thread_local_store.h

```diff
@@ -16,6 +16,145 @@
+    hist_array[0] = histograms_[1 - current_active_];
+    hist_accumulate(target, hist_array, ARRAY_SIZE(hist_array));
+    hist_clear(hist_array[0]);
+  }
+
+  Store& parent_;
+
+private:
+  uint64_t current_active_;
+  histogram_t* histograms_[2];
+  std::atomic<uint16_t> flags_;
+  HistogramStatisticsImpl interval_statistics_;
+  HistogramStatisticsImpl cumulative_statistics_;
+};
+
+typedef std::shared_ptr<ThreadLocalHistogramImpl> TlsHistogramSharedPtr;
+
+/**
+ * Log Linear Histogram implementation that is stored in the main thread.
+ */
+class HistogramParentImpl : public Histogram, public MetricImpl {
+public:
+  HistogramParentImpl(const std::string& name, Store& parent, std::string&& tag_extracted_name,
+                      std::vector<Tag>&& tags)
+      : MetricImpl(name, std::move(tag_extracted_name), std::move(tags)), parent_(parent) {
+    interval_histogram_ = hist_alloc();
+    cumulative_histogram_ = hist_alloc();
+  }
+
+  ~HistogramParentImpl() {
+    hist_free(interval_histogram_);
+    hist_free(cumulative_histogram_);
+  }
+
+  // Stats::Histogram
+  // TODO: split the Histogram interface in to two - parent and tls.
+  void recordValue(uint64_t) override { NOT_IMPLEMENTED; }
+
+  bool used() const override {
+    std::unique_lock<std::mutex> lock(merge_lock_);
+    return usedWorker();
+  }
+
+  void addTlsHistogram(TlsHistogramSharedPtr hist_ptr) {
+    std::unique_lock<std::mutex> lock(merge_lock_);
+    tls_histograms_.emplace_back(hist_ptr);
+  }
+
+  /**
+   * This method is called during the main stats flush process for each of the histogram. This
+   * method iterates through the Tls histograms and collects the histogram data of all of them
+   * in to "interval_histogram_". Then the collected "interval_histogram_" is merged to a
+   * "cumulative_histogram". More details about threading model at
+   * https://github.com/envoyproxy/envoy/issues/1965#issuecomment-376672282.
+   */
+  void merge() override {
+    std::unique_lock<std::mutex> lock(merge_lock_);
+    if (usedWorker()) {
+      hist_clear(interval_histogram_);
+      for (TlsHistogramSharedPtr tls_histogram : tls_histograms_) {
+        tls_histogram->merge(interval_histogram_);
+      }
+      histogram_t* hist_array[1];
+      hist_array[0] = interval_histogram_;
+      hist_accumulate(cumulative_histogram_, hist_array, ARRAY_SIZE(hist_array));
+      cumulative_statistics_ = HistogramStatisticsImpl(cumulative_histogram_);
+      interval_statistics_ = HistogramStatisticsImpl(interval_histogram_);
+    }
+  }
+
+  const HistogramStatistics& intervalStatistics() const override { return interval_statistics_; }
+
+  const HistogramStatistics& cumulativeStatistics() const override {
+    return cumulative_statistics_;
+  }
+
+  Store& parent_;
+  std::list<TlsHistogramSharedPtr> tls_histograms_;
+
+private:
+  bool usedWorker() const {
+    bool any_tls_used = false;
+    for (TlsHistogramSharedPtr tls_histogram : tls_histograms_) {
```

Comment:
Can you try:
```
  for (const TlsHistogramSharedPtr& tls_histogram : ...)
```
the benefit is that the shared-pointer won't have to inc/dec ref-counts during iteration, and the references are held through the loop by tls_histograms_.